3rdparty/obiwarp/obiwarp.cpp

```diff
@@ -0,0 +1,136 @@
+    // if(mzPoint)
+    //     delete[] mzPoint;
+}
+
+void ObiWarp::setReferenceData(vector<float> &rtPoints, vector<float> &mzPoints, vector<vector<float> >& intMat){
+    _tm_vals = rtPoints.size();
+    tmPoint = new float[_tm_vals];
+    for(int i=0; i < _tm_vals ; ++i)
+        tmPoint[i] = rtPoints[i];
+    _tm.take(_tm_vals, tmPoint);
+
+    _mz_vals = mzPoints.size();
+    mzPoint = new float[_mz_vals];
+    for(int i = 0; i < _mz_vals; ++i)
+        mzPoint[i] = mzPoints[i];
+    _mz.take(_mz_vals, mzPoint);
+
+    assert(_tm_vals == intMat.size());
+    MatF mat(_tm_vals, _mz_vals);
+    for(int i=0; i < _tm_vals; ++i){
+        assert(_mz_vals == intMat[i].size());
+        for(int j = 0; j < _mz_vals; ++j)
+            mat(i,j) = intMat[i][j];
+    }
+    _mat.take(mat);
+
+}
+
+vector<float> ObiWarp::align(vector<float> &rtPoints, vector<float> &mzPoints, vector<vector<float> >& intMat){
+    
+    vector<float> alignedRts;
+    if(rtPoints.size() == 0 || mzPoints.size() == 0 || _tm.len() == 0 || _mz.len() == 0)
+        return alignedRts;
+    VecF tm;
+    int tm_vals = rtPoints.size();
+    float* tmPoint = new float[tm_vals];
+    for(int i = 0; i < tm_vals; ++i)
+        tmPoint[i] = rtPoints[i];
+    tm.take(tm_vals, tmPoint);
+
+    VecF mz;
+    int mz_vals = mzPoints.size();
+    float* mzPoint = new float[mz_vals];
+    for(int i = 0; i < mz_vals; ++i)
+        mzPoint[i] = mzPoints[i];
+    mz.take(mz_vals, mzPoint);
+
+    assert(tm_vals = intMat.size());
+    MatF mat(tm_vals, mz_vals);
+    for(int i = 0; i < tm_vals; ++i){
+        assert(mz_vals == intMat[i].size());
+        for(int j = 0; j < mz_vals; ++j)
+            mat(i,j)=intMat[i][j];
+    }
+
+    MatF smat;
+    dyn.score(_mat, mat, smat, score);
+
+    if (!nostdnrm) {
+        if (!smat.all_equal()) { 
+            smat.std_normal();
+        }
+    }
+
+    int gp_length = smat.rows() + smat.cols();
+
+    VecF gp_array;
+    dyn.linear_less_before(gap_extend,gap_init,gp_length,gp_array);
+
+    int minimize = 0;
+    dyn.find_path(smat, gp_array, minimize, factor_diag, factor_gap, local, init_penalty);
+
+    VecI mOut;
+    VecI nOut;
+    dyn.warp_map(mOut, nOut, response, minimize);
+
+    VecF nOutF;
+    VecF mOutF;
+    tm_axis_vals(mOut, mOutF, _tm,_tm_vals);
+    tm_axis_vals(nOut, nOutF, tm,tm_vals); //
+    warp_tm(nOutF, mOutF, tm);
+    
+    float* rts = tm.pointer();
+    for(int i = 0; i < tm_vals; ++i)
+        alignedRts.push_back(rts[i]);
+    
+    // delete[] tmPoint;
+    // delete[] mzPoint;
+
+    return alignedRts;
+}
+
+
+void ObiWarp::tm_axis_vals(VecI &tmCoords, VecF &tmVals,VecF &_tm ,int _tm_vals){
+    VecF tmp(tmCoords.length());
+    for (int i = 0; i < tmCoords.length(); ++i) {
+        if (tmCoords[i] < _tm_vals) {
+            tmp[i] = _tm[tmCoords[i]];
+        }
+        else {
+            printf("asking for time value at index: %d (length: %d)\n", tmCoords[i], _tm_vals);
+            exit(1);
```

Comment:
So, we simply exit from here ?
We should get back to: 
1. `mzAligner` and inform that the alignment process failed
2.  tell the user that the alignment was not performed ?(can be done later ?)

Reply:
@rish9511 we'll do it when will be adding code for error handling.